doc/man/polkit-module-run-program.8.in
.\" 
.\" polkit-module-run-program manual page.
.\" Copyright (C) 2007 David Zeuthen <david@fubar.dk>
.\"
.TH POLKIT-MODULE-RUN-PROGRAM 8
.SH NAME
polkit-module-run-program \- determine policy by running a program
.SH SYNOPSIS
.PP
.HP 31
\fBpolkit-module-run-program.so\fR program=\fI<program>\fR
.SH DESCRIPTION
.PP
This PolicyKit module will run a program every time an action is
requested. For more information about the big picture refer to the
\fIPolicyKit spec\fP which can be found in
.I "@docdir@/spec/polkit-spec.html"
depending on the distribution.

.SH OPTIONS

.TP 3n
.B program=<program>
Absolute path to program to run; this is a mandatory option. Examples:
.B program=/usr/bin/my-program
or
.B program="/path/to/program --foo --bar"

.SH DESCRIPTION
This module will invoke the given program and will export a minimal
environment with values identifying the request. The program
.B SHOULD NOT
have any side effects; it is only invoked to make a decision - not to
alter state on the system. Further, the program is not guaranteed to
run as 
.B uid 0
(e.g. root); it is effectively invoked by a mechanism (such as
\fBhald\fR(7)) that may run as an unprivileged system user.

.PP
If the program fails to run or exits with a non-zero exit code, it
means that the request is denied (same as returning
.B no
- see below). If the program exits with exit code 0,
.I stdout
of the program is examined to determine the result of the decision
(these values map directly to the possible values in the
.I PolKitResult
enumeration):

.I
.TP
.B unknown
The passed action is unknown.
.TP
.B not_authorized
The mechanism / caller (e.g. the program using
.I libpolkit 
that loads this module) is not sufficiently privileged to know the
answer.
.TP
.B no
Access denied.
.TP
.B auth_root
Access denied, but authentication of the caller as root will grant
access to only that caller.
.TP
.B auth_root_keep_session
Access denied, but authentication of the caller as root will grant
access for the remainder of the session the caller stems from.
.TP
.B auth_root_keep_always
Access denied, but authentication of the caller as root will grant
access to the user of the caller in the future.
.TP
.B auth_self
Access denied, but authentication of the caller as himself will grant
access to only that caller.
.TP
.B auth_self_keep_session
Access denied, but authentication of the caller as himself will grant
access for the remainder of the session the caller stems from.
.TP
.B auth_self_keep_always
Access denied, but authentication of the caller as himself will grant
access to the user of the caller in the future.
.TP
.B yes
Access granted.

.PP
For a request concerning decisions for calls via the system message
bus daemon, the environment will contain:

.TP
.B POLKIT_REQUEST_CALLER=1
To identify the request to be concerning a decision about a caller on
the system message bus.
.TP
.B POLKIT_ACTION_ID
An identifier for the action
.TP
.B POLKIT_ACTION_<KEY_NAME>=<value-name>
All action parameters are put in the environment; the key is
uppercased and hyphen and period characters are replaced with
underscores.
.TP
.B POLKIT_RESOURCE_ID
Resource identifier
.TP
.B POLKIT_RESOURCE_TYPE
Resource type
.TP
.B POLKIT_CALLER_UID
UNIX user id of the caller
.TP
.B POLKIT_CALLER_DBUS_NAME
Unique name of the caller on the system message bus
.TP
.B POLKIT_CALLER_PID
UNIX process id of the caller
.TP
.B POLKIT_CALLER_SELINUX_CONTEXT
SELinux security context of the caller (only set if SELinux is enabled)
.TP
.B POLKIT_SESSION_CK_IS_ACTIVE
Whether ConsoleKit regards the session as active (only set if the caller belong to a session)
.TP
.B POLKIT_SESSION_CK_IS_LOCAL
Whether ConsoleKit regards the session as local (only set if the caller belong to a session)
.TP
.B POLKIT_SESSION_CK_OBJREF
ConsoleKit session D-Bus object reference (only set if the caller belong to a session)
.TP
.B POLKIT_SESSION_UID
UNIX user ID of the user owning the session (only set if the caller belong to a session)
.TP
.B POLKIT_SEAT_CK_OBJREF
ConsoleKit seat D-Bus object reference of the seat that the session belongs to (only set if the caller belong to a session)

.PP
For a request concerning session-wide decisions the environment will
contain:

.TP
.B POLKIT_REQUEST_SESSION=1
To identify the request to be session-wide.
.TP
.B POLKIT_ACTION_ID
An identifier for the action
.TP
.B POLKIT_RESOURCE_ID
Resource identifier
.TP
.B POLKIT_RESOURCE_TYPE
Resource type
.TP
.B POLKIT_SESSION_CK_IS_ACTIVE
Whether ConsoleKit regards the session as active
.TP
.B POLKIT_SESSION_CK_IS_LOCAL
Whether ConsoleKit regards the session as local
.TP
.B POLKIT_SESSION_CK_OBJREF
ConsoleKit session D-Bus object reference
.TP
.B POLKIT_SESSION_UID
UNIX user ID of the user owning the session
.TP
.B POLKIT_SEAT_CK_OBJREF
ConsoleKit seat D-Bus object reference of the seat that the session belongs to

.SH NOTES
.PP
As PolicyKit modules are heavily used to enforce policy, running a
program on every request may put unneccessary load on the system
unless judicious use of built-in options to confine the module are
employed.

.SH BUGS
.PP
Please send bug reports to either the distribution or the HAL
mailing list, see 
.I "http://lists.freedesktop.org/mailman/listinfo/hal"
on how to subscribe.

.SH SEE ALSO
.PP
\&\fIPolicyKit\fR\|(8),
\&\fIhald\fR\|(8),
\&\fIdbus-daemon\fR\|(1),
\&\fIpolkit-module-builtins\fR\|(8),
\&\fIpolkit-module-default\fR\|(8),
\&\fIpolkit-module-allow-all\fR\|(8),
\&\fIpolkit-module-deny-all\fR\|(8),
\&\fI@sysconfdir@/PolicyKit/policy\fR\|,
\&\fI@sysconfdir@/PolicyKit/PolicyKit.conf\fR\|

.SH AUTHOR
Written by David Zeuthen <david@fubar.dk> with a lot of help from many
others.


NEWS
==========
PolicyKit 0.1 ""
==========

WRITE ME

Requirements for PolicyKit 0.1 "" (and git master)

 - glib             >= 2.6.0


modules/run-program/polkit-module-run-program.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*- */
/***************************************************************************
 *
 * polkit-module-run-program.c : determine policy by running a program
 *
 * Copyright (C) 2007 David Zeuthen, <david@fubar.dk>
 *
 * Licensed under the Academic Free License version 2.1
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307	 USA
 *
 **************************************************************************/

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <libpolkit/libpolkit.h>
#include <glib.h>

/* The symbol that libpolkit looks up when loading this module */
polkit_bool_t libpolkit_module_set_functions (PolKitModuleInterface *module_interface);

typedef struct {
        int program_argc;
        char **program_argv;
} UserData;

static polkit_bool_t
_module_init (PolKitModuleInterface *module_interface, int argc, char *argv[])
{
        int n;
        UserData *user_data;

        user_data = g_new0 (UserData, 1);
        for (n = 1; n < argc; n++) {
                if (g_str_has_prefix (argv[n], "program=")) {
                        const char *program;
                        program = argv[n] + 8;

                        if (!g_shell_parse_argv (program, 
                                                 &user_data->program_argc, 
                                                 &user_data->program_argv, NULL)) {
                                g_warning ("Cannot parse '%s' - skipping", program);
                                goto error;
                        }

                        if (!g_file_test (user_data->program_argv[0], 
                                          G_FILE_TEST_IS_EXECUTABLE|G_FILE_TEST_IS_REGULAR)) {
                                g_warning ("Program '%s' is not an executable file - skipping",
                                           user_data->program_argv[0]);
                                goto error;
                        }

                        g_debug ("program = '%s'", user_data->program_argv[0]);

                        /* TODO:
                         * O_o o_O... we could monitor the executable file :-) and trigger config changes!
                         */
                }
        }

        if (user_data->program_argv == NULL)
                goto error;

        libpolkit_module_set_user_data (module_interface, user_data);

        return TRUE;
error:
        if (user_data->program_argv != NULL)
                g_strfreev (user_data->program_argv);
        g_free (user_data);
        return FALSE;
}

static void
_module_shutdown (PolKitModuleInterface *module_interface)
{
        UserData *user_data;
        user_data = libpolkit_module_get_user_data (module_interface);
        if (user_data != NULL) {
                if (user_data->program_argv != NULL)
                        g_strfreev (user_data->program_argv);
                g_free (user_data);
        }
}

static polkit_bool_t
_add_action_to_env (PolKitAction *action, GPtrArray *envp)
{
        char *p_id;
        if (!libpolkit_action_get_action_id (action, &p_id))
                goto error;
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_ACTION_ID=%s", p_id));
        return TRUE;
error:
        return FALSE;
}

static polkit_bool_t
_add_resource_to_env (PolKitResource *resource, GPtrArray *envp)
{
        char *r_type;
        char *r_id;
        if (!libpolkit_resource_get_resource_type (resource, &r_type))
                goto error;
        if (!libpolkit_resource_get_resource_id (resource, &r_id))
                goto error;
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_RESOURCE_TYPE=%s", r_type));
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_RESOURCE_ID=%s", r_id));
        return TRUE;
error:
        return FALSE;
}

static polkit_bool_t
_add_seat_to_env (PolKitSeat *seat, GPtrArray *envp)
{
        char *s_ck_objref;
        if (!libpolkit_seat_get_ck_objref (seat, &s_ck_objref))
                goto error;
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_SEAT_CK_OBJREF=%s", s_ck_objref));
        return TRUE;
error:
        return FALSE;
}

static polkit_bool_t
_add_session_to_env (PolKitSession *session, GPtrArray *envp)
{
        uid_t s_uid;
        char *s_ck_objref;
        polkit_bool_t s_ck_is_active;
        polkit_bool_t s_ck_is_local;
        char *s_ck_remote_host;
        PolKitSeat *s_seat;

        if (!libpolkit_session_get_uid (session, &s_uid))
                goto error;
        if (!libpolkit_session_get_ck_objref (session, &s_ck_objref))
                goto error;
        if (!libpolkit_session_get_ck_is_active (session, &s_ck_is_active))
                goto error;
        if (!libpolkit_session_get_ck_is_local (session, &s_ck_is_local))
                goto error;
        if (!s_ck_is_local)
                if (!libpolkit_session_get_ck_remote_host (session, &s_ck_remote_host))
                        goto error;
        if (!libpolkit_session_get_seat (session, &s_seat))
                goto error;

        if (!_add_seat_to_env (s_seat, envp))
                goto error;
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_SESSION_UID=%d", (int) s_uid));
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_SESSION_CK_OBJREF=%s", s_ck_objref));
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_SESSION_CK_IS_ACTIVE=%d", s_ck_is_active));
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_SESSION_CK_IS_LOCAL=%d", s_ck_is_local));
        if (!s_ck_is_local)
                g_ptr_array_add (envp, g_strdup_printf ("POLKIT_SESSION_CK_REMOTE_HOST=%s", s_ck_remote_host));
        return TRUE;
error:
        return FALSE;
}

static polkit_bool_t
_add_caller_to_env (PolKitCaller *caller, GPtrArray *envp)
{
        uid_t c_uid;
        pid_t c_pid;
        char *c_selinux_context;
        char *c_dbus_name;
        PolKitSession *c_session;

        if (!libpolkit_caller_get_uid (caller, &c_uid))
                goto error;
        if (!libpolkit_caller_get_pid (caller, &c_pid))
                goto error;
        if (!libpolkit_caller_get_dbus_name (caller, &c_dbus_name))
                goto error;
        if (!libpolkit_caller_get_selinux_context (caller, &c_selinux_context)) /* SELinux may not be available */
                c_selinux_context = NULL;
        if (!libpolkit_caller_get_ck_session (caller, &c_session)) /* Caller may not originate from a session */
                c_session = NULL;

        if (c_session != NULL)
                if (!_add_session_to_env (c_session, envp))
                        goto error;
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_CALLER_UID=%d", (int) c_uid));
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_CALLER_PID=%d", (int) c_pid));
        g_ptr_array_add (envp, g_strdup_printf ("POLKIT_CALLER_DBUS_NAME=%s", c_dbus_name));
        if (c_selinux_context != NULL)
                g_ptr_array_add (envp, g_strdup_printf ("POLKIT_CALLER_SELINUX_CONTEXT=%s", c_selinux_context));
        return TRUE;
error:
        return FALSE;
}

static polkit_bool_t
_run_program (UserData *user_data, char **envp, PolKitResult *result)
{
        int n;
        int exit_status;
        GError *g_error;
        char *prog_stdout;
        polkit_bool_t ret;

        g_error = NULL;
        prog_stdout = NULL;
        ret = FALSE;

        if (!g_spawn_sync ("/",
                           user_data->program_argv,
                           envp,
                           0,
                           NULL,
                           NULL,
                           &prog_stdout,
                           NULL,
                           &exit_status,
                           &g_error)) {
                g_warning ("error spawning '%s': %s", user_data->program_argv[0], g_error->message);
                g_error_free (g_error);
                goto error;
        }

        /* only care if the program returned 0 */
        if (exit_status != 0)
                goto error;

        /* only care about the first line */
        for (n = 0; prog_stdout[n] != '\n' && prog_stdout[n] != '\0'; n++)
                ;
        prog_stdout[n] = '\0';

        if (!libpolkit_result_from_string_representation (prog_stdout, result)) {
                g_warning ("malformed result '%s' from program", prog_stdout);
                goto error;
        }

        ret = TRUE;
error:
        g_free (prog_stdout);
        return ret;
}


static PolKitResult
_module_can_session_access_resource (PolKitModuleInterface *module_interface,
                                     PolKitContext         *pk_context,
                                     PolKitAction          *action,
                                     PolKitResource        *resource,
                                     PolKitSession         *session)
{
        PolKitResult result;
        UserData *user_data;
        GPtrArray *envp;

        envp = NULL;
        result = LIBPOLKIT_RESULT_UNKNOWN_ACTION;

        user_data = libpolkit_module_get_user_data (module_interface);

        envp = g_ptr_array_new ();

        if (!_add_action_to_env (action, envp))
                goto error;
        if (resource != NULL)
                if (!_add_resource_to_env (resource, envp))
                        goto error;
        if (!_add_session_to_env (session, envp))
                goto error;
        g_ptr_array_add (envp, g_strdup ("PATH=/usr/bin:/bin"));
        g_ptr_array_add (envp, g_strdup ("POLKIT_REQUEST_SESSION=1"));
        g_ptr_array_add (envp, NULL);

        if (!_run_program (user_data, (char **) envp->pdata, &result))
                goto error;
        
error:
        if (envp != NULL) {
                g_ptr_array_foreach (envp, (GFunc) g_free, NULL);
                g_ptr_array_free (envp, TRUE);
        }
        return result;
}

static PolKitResult
_module_can_caller_access_resource (PolKitModuleInterface *module_interface,
                                    PolKitContext         *pk_context,
                                    PolKitAction          *action,
                                    PolKitResource        *resource,
                                    PolKitCaller          *caller)
{
        PolKitResult result;
        UserData *user_data;
        GPtrArray *envp;

        envp = NULL;
        result = LIBPOLKIT_RESULT_NO;
        user_data = libpolkit_module_get_user_data (module_interface);

        envp = g_ptr_array_new ();
        if (!_add_action_to_env (action, envp))
                goto error;
        if (resource != NULL)
                if (!_add_resource_to_env (resource, envp))
                        goto error;
        if (!_add_caller_to_env (caller, envp))
                goto error;
        g_ptr_array_add (envp, g_strdup ("PATH=/usr/bin:/bin"));
        g_ptr_array_add (envp, g_strdup ("POLKIT_REQUEST_CALLER=1"));
        g_ptr_array_add (envp, NULL);
        if(!_run_program (user_data, (char **) envp->pdata, &result))
                goto error;

error:
        if (envp != NULL) {
                g_ptr_array_foreach (envp, (GFunc) g_free, NULL);
                g_ptr_array_free (envp, TRUE);
        }
        return result;
}

polkit_bool_t
libpolkit_module_set_functions (PolKitModuleInterface *module_interface)
{
        polkit_bool_t ret;

        ret = FALSE;
        if (module_interface == NULL)
                goto out;

        libpolkit_module_set_func_initialize (module_interface, _module_init);
        libpolkit_module_set_func_shutdown (module_interface, _module_shutdown);
        libpolkit_module_set_func_can_session_access_resource (module_interface, _module_can_session_access_resource);
        libpolkit_module_set_func_can_caller_access_resource (module_interface, _module_can_caller_access_resource);

        ret = TRUE;
out:
        return ret;
}
